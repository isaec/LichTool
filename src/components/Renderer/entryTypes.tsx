import { Component, createMemo, For, JSX, Show } from "solid-js";
import { DataGroupRenderer, DataRenderer } from "./Renderer";
import {
  BonusData,
  InsetData,
  isDataNode,
  ListData,
  listStyles,
  QuoteData,
  SectionData,
  BonusSpeed,
} from "./types";

import styles from "./Renderer.module.scss";

const ListItem: Component<{ condition: boolean; children: JSX.Element }> = (
  props
) => (props.condition ? props.children : <li>{props.children}</li>);

const entryTypes = new Map(
  Object.entries({
    // each key corresponds to a "type": "xxx"
    // each value is jsx to render that type
    section: (props: { data: SectionData }) => (
      <>
        <h1>{props.data.name}</h1>
        <DataGroupRenderer group={props.data.entries} />
      </>
    ),
    list: (props: { data: ListData }) => {
      const classes = createMemo(() => {
        // if there is no style, we don't need to give it a class
        if (props.data.style === undefined) return "";
        // if there is only one style, give it the appropriate class
        if (listStyles.has(props.data.style)) return styles[props.data.style];

        // otherwise, build up a string of each converted style
        const classStack: Array<string> = [];
        listStyles.forEach((style) => {
          if (props.data.style!.includes(style)) classStack.push(styles[style]);
        });
        return classStack.join(" ");
      });
      return (
        <>
          <Show when={props.data.name !== undefined}>
            <b>{props.data.name}</b>
          </Show>
          <ul
            classList={{
              [styles.list]: true,
              [styles.column]: typeof props.data.columns === "number",
              [classes()]: true,
            }}
            style={
              typeof props.data.columns === "number"
                ? {
                    "column-count": props.data.columns,
                  }
                : undefined
            }
          >
            <For each={props.data.items}>
              {(item) => (
                <ListItem condition={isDataNode(item) && item.type === "list"}>
                  <DataRenderer data={item} />
                </ListItem>
              )}
            </For>
          </ul>
        </>
      );
    },
    inset: (props: { data: InsetData }) => (
      <div class={styles.inset}>
        <h4>{props.data.name}</h4>
        {<DataRenderer data={props.data.entries} />}
      </div>
    ),
    quote: (props: { data: QuoteData }) => {
      const entries = createMemo(() => {
        // if we only have one string, quote it
        if (
          props.data.entries.length === 1 &&
          typeof props.data.entries[0] === "string"
        ) {
          return [`"${props.data.entries}"`];
        }

        // we have many things, but some elements may be strings
        const arr = props.data.entries.slice(1, -1);

        if (typeof props.data.entries[0] === "string") {
          arr.unshift(`"${props.data.entries[0]}`);
        } else {
          arr.unshift(`"`);
          arr.unshift(props.data.entries[0]); // also add the missing non string element
        }

        if (
          typeof props.data.entries[props.data.entries.length - 1] === "string"
        ) {
          arr.push(`${props.data.entries[props.data.entries.length - 1]}"`);
        } else {
          arr.push(props.data.entries[props.data.entries.length - 1]); // add the missing non string element
          arr.push(`"`);
        }

        return arr;
      });
      return (
        <figure>
          <DataGroupRenderer group={entries()} wrapper={"blockquote"} />
          <Show when={typeof (props.data.by ?? props.data.from) !== undefined}>
            <figcaption>
              â€” {props.data.by}
              <Show
                when={
                  typeof props.data.by === "string" &&
                  typeof props.data.from === "string"
                }
              >
                ,{" "}
              </Show>
              <cite>{props.data.from}</cite>
            </figcaption>
          </Show>
        </figure>
      );
    },
    bonus: (props: { data: BonusData }) => (
      <span>
        <Show when={props.data.value >= 0}>+</Show>
        {props.data.value}
      </span>
    ),
    bonusSpeed: (props: { data: BonusSpeed }) => (
      <span>
        <Show when={props.data.value >= 0}>+</Show>
        {props.data.value}
        {" ft."}
      </span>
    ),
  })
);
export default entryTypes;
