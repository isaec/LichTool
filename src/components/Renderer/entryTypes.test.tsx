import { Component } from "solid-js";
import { Dynamic } from "solid-js/web";
import { render } from "solid-testing-library";
import { vi, expect, describe, it, test } from "vitest";
import { generate, one, optional, illegal } from "generate-combinations";
import entryTypes from "./entryTypes";
import styles from "./Renderer.module.scss";
import {
  BonusData,
  BonusSpeed,
  DataNode,
  InsetData,
  InsetReadaloudData,
  ListData,
  QuoteData,
  SectionData,
} from "./types";

vi.mock("./Renderer.module.scss", () => ({
  default: new Proxy(new Object(), {
    get(_, style) {
      return style;
    },
  }),
}));

describe("entryTypes", () => {
  // this is used to test how embedded type behaves
  const inset: InsetData = { type: "inset", name: "name", entries: ["inset!"] };
  const tests: Array<DataNode> = [
    ...generate<SectionData>({
      type: "section",
      name: "name",
      entries: one([["entry"], ["entry", inset]]),
    }),
    ...generate<ListData>({
      type: "list",
      items: one([["stuff"], ["stuff", inset]]),
    }),
    ...generate<InsetData | InsetReadaloudData>({
      type: illegal<any, any>(one(["inset", "insetReadaloud"])),
      name: one(["name"]),
      entries: one([["woo"], ["a", "b", "c"]]),
    }),
    ...generate<QuoteData>({
      type: "quote",
      entries: one([
        ["body of the quote"],
        ["two entries", "for this quote"],
        ["quote with an", inset],
      ]),
      by: optional("author"),
      from: optional("source"),
    }),
    ...generate<BonusData>({
      type: "bonus",
      value: one([1, -2, 0]),
    }),
    ...generate<BonusSpeed>({
      type: "bonusSpeed",
      value: one([100, -2, 15]),
    }),
  ];
  it.each(tests)(`rendering %s matches snapshot`, (data) => {
    const { unmount, container } = render(() => (
      <Dynamic component={entryTypes.get(data.type) as any} data={data} />
    ));

    expect(container).toMatchSnapshot();
    unmount();
  });
});
