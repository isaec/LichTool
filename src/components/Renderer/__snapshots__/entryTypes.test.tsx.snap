// Vitest Snapshot v1

exports[`entryTypes > rendering {
  type: 'quote',
  entries: [Array],
  by: 'Anon',
  from: 'Archive of Lost Chats'
} matches snapshot 1`] = `
<div>
  <figure>
    <blockquote>
      <p>
        "Look, don't quote me on this, but"
      </p>
    </blockquote>
    <figcaption>
      â€” 
      Anon
      ,
       
      <cite>
        Archive of Lost Chats
      </cite>
    </figcaption>
  </figure>
</div>
`;
